centrallix-sysdoc/OSDriver_Authoring.md

```diff
@@ -115,497 +170,1032 @@ License:  Copyright (C) 2001-2011 LightSys Technology Services. See LICENSE.txt
+It is common practice to use `stLookup()` and `stAttrValue()` or `stGetExpression()` (see below) together to retrieve values, for example (where `inf` is a `pStructInfo` variable from somewhere):
 
-### int stFreeInf(pStructInf this)
-This function is used to free a StructInf tree node.  It will free any sub-nodes first, so if that is not desired, be sure to disconnect them by removing them from the SubInf array and appropriately adjusting the nSubInf counter, and setting the SubInf array position to NULL.  This function also disconnects the tree node from its parent, if any, so if the parent is already free()'d, be sure to set the node's Parent pointer to NULL.  Any strings marked allocated with the StrAlloc flags will be free()'d.
+```c
+char* ptr;
+if (stAttrValue(stLookup(inf, "my_attr"), NULL, &ptr, 0) != 0)
+    goto error_handling;
+printf("The value is: %s\n", ptr);
+```
 
-It is also common practice to bypass the stXxx() functions entirely and access the elements of the StructInf structures themselves.  This is not forbidden, and may be done.  See the file stparse.h for a description of the structure.  For example,
 
-    pStructInf inf;
-    int i;
+### stparse: stGetExpression()
+```c
+pExpression stGetExpression(pStructInf this, int nval);
+```
+Returns a pointer to an expression that represents the value of the nval-th element of the given struct.
+
+
+### stparse: stCreateStruct()
+```c
+pStructInf stCreateStruct(char* name, char* type);
+```
+This function creates a new top-level tree item of type `ST_T_STRUCT`, with a given name and content-type.
+
+
+### stparse: stAddAttr()
+```c
+pStructInf stAddAttr(pStructInf inf, char* name);
+```
+This function adds a node of type `ST_T_ATTRIB` to either an `ST_T_STRUCT` or an `ST_T_SUBGROUP` type of node, with a given name and no values (see AddValue, below).  The new attribute tree node is linked under the `inf` node passed, and is returned.
+
+
+### stparse: stAddGroup()
+```c
+pStructInf stAddGroup(pStructInf inf, char* name, char* type);
+```
+This function adds a node of type `ST_T_SUBGROUP` to either an `ST_T_SUBGROUP` or an `ST_T_STRUCT` tree node, with a given name and content type (content type such as `"report/query"`).
+
+
+### stparse: stAddValue()
+```c
+int stAddValue(pStructInf inf, char* strval, int intval);
+```
+This function adds a value to an attribute, and can be called multiple times on an attribute to add a list of values.  If `strval` is not null, a string value is added, otherwise an integer value is added.  The string is NOT copied, but is simply pointed-to.  If the string is non-static, and has a lifetime less than the `ST_T_ATTRIB` tree node, then the following procedure should be used, where `str` is the string pointer to the string:
+
+```c
+pStructInf attr_inf = stAddAttr(my_parent_inf, "my_attr");
+if (attr_inf == NULL) goto error_handling;
+
+char* new_str = (char*)malloc(strlen(str) + 1lu);
+if (new_str == NULL) goto error_handling;
+strcpy(new_str, str);
```

Comment:
This pattern is for the stAddValue_ne() call which operates on a pStruct instead of pStructInf.  That's a whole different thing to document.  For stAddValue(), the string is properly copied/allocated in the StructInf structure.

Reply:
Are you sure this pattern is not necessary? `stAddValue()`'s implementation does not appear to copy the string.

```c
int
stAddValue(pStructInf inf, char* strval, int intval)
    {

	if (inf->nVal == ST_MAX_VALUES) return 0;
	if (strval)
	    {
	    inf->StrVal[inf->nVal] = strval;
	    inf->IntVal[inf->nVal++] = 0;
	    }
	else
	    {
	    inf->StrVal[inf->nVal] = NULL;
	    inf->IntVal[inf->nVal++] = intval;
	    }

    return inf->nVal;
    }
```

Reply:
Wait, I didn't realize there were multiple. It's confusing my editor, and it keeps jumping to the wrong function, causing confusion. Let me try to mess with the settings and see if I can fix it.